### PR TITLE
Active Scan Analyser will now be used more often

### PR DESCRIPTION
**Pull request by kingthorin**
The purpose of this change is so that analyser will be used based on more potential starting points, therefore giving us better coverage for deciding on 404s etc when falling back from custom pages (ex: custom pages aren't configured).

(Using diff highlighting below to illustrate coverage)
Before:
```diff
+ http://demo.testfire.net - Yes (Analyser ran)
-     GET:/ - No (Analyzer didn't run)
+     images - Yes (Analyser ran)
-        GET:gradient.jpg  - No (Analyzer didn't run)
-     GET:style.css - No (Analyzer didn't run)
```

After:
```diff
+ http://demo.testfire.net - Yes (Analyser ran)
+     GET:/ - Yes (Analyser ran)
+     images - Yes (Analyser ran)
+        GET:gradient.jpg  - Yes (Analyser ran)
+     GET:style.css - Yes (Analyser ran)
```

Related to: zaproxy/zaproxy#7036 and discussed via IRC.

To enable debug logging for Analyser:
```js
var Configurator = Java.type("org.apache.logging.log4j.core.config.Configurator");
var Level = Java.type("org.apache.logging.log4j.Level");

// https://logging.apache.org/log4j/2.x/javadoc.html
Configurator.setLevel({
  "org.parosproxy.paros.core.scanner.Analyser" : Level.DEBUG,
});
```

Signed-off-by: kingthorin <kingthorin@users.noreply.github.com>

### CODE REVIEW

**Comment by kingthorin** [comment]
Feel free to edit the subject/title as you see fit to make a better change notes entry :grinning:

File: zap/src/main/java/org/parosproxy/paros/core/scanner/Analyser.java [hunk 1]
```diff
@@ -364,7 +364,8 @@ private int inOrderAnalyse(StructuralNode node) {
             if (!node.isRoot()) {
                 if (!node.isLeaf() || node.isLeaf() && node.getParent().isRoot()) {
                     analyse(node);
-
+                } else if (node.isLeaf() && !node.getParent().isRoot()) {
```
Comment (psiinon) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/Analyser.java:367]: Will need a `// ZAP: ` comment?
Would it be possible / practical to add unit tests?

Comment (psiinon) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/Analyser.java:367]: Could always change the visibility of this method to make that easier...

Comment (thc202) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/Analyser.java:367]: Better add tests to verify that the leafs are handled just once.

Comment (kingthorin) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/Analyser.java:367]: I'm happy to add tests but I'll need an example here too.

Comment (kingthorin) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/Analyser.java:366]: I started to try to figure something out here, but got stuck:
<details>
<summary>Code: Attempted Initial UnitTest</summary>

```java
package org.parosproxy.paros.core.scanner;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import org.apache.commons.httpclient.URIException;
import org.junit.jupiter.api.Test;
import org.parosproxy.paros.db.DatabaseException;
import org.parosproxy.paros.db.RecordHistory;
import org.parosproxy.paros.db.TableHistory;
import org.parosproxy.paros.model.HistoryReference;
import org.parosproxy.paros.model.Model;
import org.parosproxy.paros.model.SiteMap;
import org.parosproxy.paros.model.SiteNode;
import org.parosproxy.paros.network.HttpMalformedHeaderException;
import org.zaproxy.zap.model.StructuralNode;
import org.zaproxy.zap.model.StructuralSiteNode;

class AnalyserUnitTest {

    @SuppressWarnings("deprecation")
    @Test
    void shouldTestSomething() throws URIException, HttpMalformedHeaderException, DatabaseException {
        // Given
        RecordHistory recHist = mock(RecordHistory.class);
        TableHistory tableHist = mock(TableHistory.class);
        when(tableHist.read(1)).thenReturn(recHist);
        SiteNode rootNode = mock(SiteNode.class);
        Model model = mock(Model.class);
        SiteMap map = new SiteMap(rootNode, model);
        HistoryReference hr = new HistoryReference(1);
        hr.setSiteNode(new SiteNode(map, HistoryReference.TYPE_ZAP_USER, "https://example.org"));
        map.addPath(hr);
        hr = new HistoryReference(2);
        hr.setSiteNode(new SiteNode(map, HistoryReference.TYPE_ZAP_USER, "https://example.org/"));
        map.addPath(hr);
        hr = new HistoryReference(3);
        hr.setSiteNode(new SiteNode(map, HistoryReference.TYPE_ZAP_USER, "https://example.org/foo.html"));
        map.addPath(hr);
        hr = new HistoryReference(4);
        hr.setSiteNode(new SiteNode(map, HistoryReference.TYPE_ZAP_USER, "https://example.org/images/"));
        map.addPath(hr);
        hr = new HistoryReference(5);
        hr.setSiteNode(new SiteNode(map, HistoryReference.TYPE_ZAP_USER, "https://example.org/images/image.jpg"));
        map.addPath(hr);
        SiteNode aNode = new SiteNode(map, HistoryReference.TYPE_ZAP_USER, "https://example.org");
        StructuralNode node = new StructuralSiteNode(aNode);
        Analyser analyser = new Analyser();
        // When
        int result = analyser.inOrderAnalyse(node);
        // Then
        System.out.println(result);
    }
}
```

</details>

It NPEs at `HistoryReference hr = new HistoryReference(1);` when doing `history = staticTableHistory.read(historyId);` in HistoryReference.

<details>
<summary>Stack Trace</summary>

```trace
java.lang.NullPointerException
	at org.parosproxy.paros.model.HistoryReference.<init>(HistoryReference.java:352)
	at org.parosproxy.paros.model.HistoryReference.<init>(HistoryReference.java:326)
	at org.parosproxy.paros.core.scanner.AnalyserUnitTest.shouldTestSomething(AnalyserUnitTest.java:31)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:95)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:91)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:60)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:98)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
```

</details>

**Comment by kingthorin** [comment]
The current set of tests aren’t quite right yet. I’ve got some of the given/when wrong. Will work on it more tomorrow and the next day.

**Comment by kingthorin** [comment]
Proving that nothing is ever simple the original code sometimes provides misleading output from `inOrderAnalyse` to start with :cry:

**Comment by kingthorin** [comment]
As discussed I've removed the UnitTest(s) and implemented a bunch of debug logging. I have yet to run it through live testing to ensure it behaves as wanted/expected.

**Comment by kingthorin** [comment]
As of [8c7a784](https://github.com/zaproxy/zaproxy/pull/7181/commits/8c7a784189c0e42348c952d4c598e218baf423e9) the coverage for a site node (ex: https://demo.testfire.net/) is the same, though slightly more chatty with the new logging. The results are the same Analyser runs both on the site and the /images directory. However, running on the `GET: /` node does not produce the same result as the site node, and likely should. I believe this is because in the new condition node should be set to `node = node.getParent()` so that then the later child iteration behaves properly. However that means that `inOrderAnalysis` needs to track what’s been analyzed more globally to avoid an endless loop of repeatedly getting the same parent.

**Comment by kingthorin** [comment]
I believe this is now working as expected.

**Comment by kingthorin** [comment]
Rebased current, and tested with bodgeitns (which uses `page` as a structural parameter). Still behaves as I believe we want it to.

**Comment by kingthorin** [comment]
Rebased and addressed conflict.

File: zap/src/main/java/org/parosproxy/paros/core/scanner/Analyser.java (hunk 1 above)
Comment (ricekot) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/Analyser.java:380]: Haven't looked at it in detail but you could try doing
```java
HistoryReference.setTableHistory(new ParosTableHistory());
```
at the beginning of the test.

Comment (kingthorin) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/Analyser.java:380]: Thanks for the idea but then I get stuck when it tries to read, or if I mock it then it fails in <init>..

**Comment by kingthorin** [comment]
> Rebased current, and tested with bodgeitns (which uses page as a structural parameter). Still behaves as I believe we want it to.

With regard to structural parameters, there is further improvement we could make. Assuming there's a somewhat easy way to identify them we could try to establish 404 handling for the relevant param/value there as well. Note: This was also a miss prior to this change, therefore I suggest it be done in separate PR (to keep this one focused/moving).

**Comment by kingthorin** [comment]
Rebased current and addressed comments received via IRC.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by psiinon — APPROVED** [review]

**Pull request merged**